python/ray/actor.py

```diff
@@ -431,7 +431,16 @@ def _ray_from_modified_class(
         # Make sure the actor class we are constructing inherits from the
         # original class so it retains all class properties.
         class DerivedActorClass(cls, modified_class):
-            pass
+            def __init__(self, *args, **kwargs):
+                if (
+                    isinstance(args[-1], dict)
+                    and "__module__" in args[-1]
+                    and "__qualname__" in args[-1]
+                    and "__init__" in args[-1]
+                ):
+                    cls.__init__(self, *args, **kwargs)
+                else:
+                    modified_class.__init__(self, *args, **kwargs)
```

Comment:
**Problem** - We cannot always delegate call to `cls.__init__` or `modified_cls.__init__`. Because if always delegate call to `cls.__init__` from here, then user defined class's `__init__` method will be ignore leading to issues like, #21868. If we always delegate call to `modified_cls.__init__` then it will allow inheriting from actor classes leading to failure of `test_actor_inheritance`. So, I have added this `if-else` check to figure out which `__init__` method should be called. If `"__module__"`, `"__qualname__"` and `"__init__"` are present in `args[-1]` then it would mean an actor class is being inherited so `cls.__init__` should be called. However, if no such signal is received in `args` then user defined class's `__init__` i.e., `modified_class.__init__` should be called.
cc: @pcmoritz

Reply:
The fix is not clear to me. Can you clarify what args is in this context? Also, why is an actor class being inherited if "__module__", "__qualname__" and "__init__" are in args[-1]?

Can you replace the test with one that is clearer?

Reply:
> Also, why is an actor class being inherited if "**module**", "**qualname**" and "**init**" are in args[-1]?

This is a pattern that I noticed. When I inherited a class, then there were 3 arguments, ("<DerivedClassName>", "<ModuleOfTheDerivedClass>",  "<A dict having module, qualname and init>"). If you notice the signature of `ActorClass.__init__`, then it matches with these 3 arguments. However, when you simply call `cls(<arguments>)` (as done in #21868), then the above pattern is not observed which means the `ActorClass` is not inherited. I know this is somewhat hackish, but let me know if Python has any other technique to figure if some class can be inherited or not.

The other approaches I tried included, checking `DerivedActorClass.__subclasses__` but that doesn't work. I am not sure if there is any other way to figure out if the user is inheriting an actor class or simply calling the `__init__` method from a `classmethod`.